### PR TITLE
feature: Add framework, framework-version, container-type arguments to deep_learning_container.py

### PR DESCRIPTION
**Pull request by TusharKanekiDey**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: deep_learning_container.py [hunk 1]
```diff
@@ -69,17 +67,41 @@ def _retrieve_instance_region():
...
     response = None
     instance_id = _retrieve_instance_id()
     region = _retrieve_instance_region()
+    ARGS = parse_args()
+    framework, framework_version, py_version = ARGS.framework, ARGS.framework_version, ARGS.python_version
 
     if instance_id is not None and region is not None:
         url = ("https://aws-deep-learning-containers-{0}.s3.{0}.amazonaws.com"
-               "/dlc-containers.txt?x-instance-id={1}".format(region, instance_id))
+               "/dlc-containers.txt?x-instance-id={1},framework={2},framework_version={3},py_version={4}".format(region, instance_id, framework, framework_version, py_version))
```
Comment (arjkesh) [inline deep_learning_container.py]: i see that these three args are not required. if they are not specified, what does the request look like? does this affect the BI team's logic for parsing the logs?

**Comment by arjkesh** [comment]
Have you tried building a container with these changes?

File: deep_learning_container.py (hunk 1 above)
Comment (TusharKanekiDey) [reply deep_learning_container.py]: Valid point. I have made the required change.

**Comment by TusharKanekiDey** [comment]
@arjkesh Yes I have build and test the container. I have also add the test_mode.

File: deep_learning_container.py [hunk 2]
```diff
@@ -69,18 +69,52 @@ def _retrieve_instance_region():
     return region
 
 
+def parse_args():
+    """Parsing function to parse input arguments.
```
Comment (arjkesh) [inline deep_learning_container.py]: nit: follow style convention of rest of file in comments, example

```
"""
Parse input arguments

return: parsed arguments object
"""
```

File: deep_learning_container.py [hunk 3]
```diff
@@ -69,18 +69,52 @@ def _retrieve_instance_region():
...
+
+
 def query_bucket():
     """
     GET request on an empty object from an Amazon S3 bucket
     """
     response = None
     instance_id = _retrieve_instance_id()
     region = _retrieve_instance_region()
+    ARGS = parse_args()
```
Comment (arjkesh) [inline deep_learning_container.py]: any reason this needs to be all caps? it's not a global right?

File: deep_learning_container.py [hunk 4]
```diff
@@ -69,18 +69,52 @@ def _retrieve_instance_region():
...
+    framework, framework_version, job = ARGS.framework, ARGS.framework_version, ARGS.job
+    py_version = sys.version[0:5]
```
Comment (arjkesh) [inline deep_learning_container.py]: I think there could be an example where this doesn't work out as expected. I.e. python 2.7.10 would end up being 2.7.1. Instead you can try to parse sys.version as so:

`sys.version.split(" ")[0]`

Comment (arjkesh) [reply deep_learning_container.py]: additionally, sys.version_info gives you (major=3, minor=7, micro=4, releaselevel='final', serial=0)

File: deep_learning_container.py [hunk 5]
```diff
@@ -69,18 +69,52 @@ def _retrieve_instance_region():
...
 
     if instance_id is not None and region is not None:
         url = ("https://aws-deep-learning-containers-{0}.s3.{0}.amazonaws.com"
-               "/dlc-containers.txt?x-instance-id={1}".format(region, instance_id))
+               "/dlc-containers.txt?x-instance-id={1},framework={2},framework_version={3},py_version={4},job={5}".format(region, instance_id, framework, framework_version, py_version, job))
         response = requests_helper(url, timeout=0.2)
+        if (os.environ.get('TEST_MODE') == str(1)):
+            f = open('/tmp/test_request.txt', "w+")
```
Comment (arjkesh) [inline deep_learning_container.py]: please use context manager for file operations, and os.path.join for file paths. With the context manager, you won't have to worry about calling f.close()

```
with open(os.path.join(os.sep, 'tmp', 'test_request.txt'), 'w+') as rf:
    rf.write(url)
```

File: deep_learning_container.py [hunk 6]
```diff
@@ -69,18 +69,52 @@ def _retrieve_instance_region():
...
+def parse_args():
+    """Parsing function to parse input arguments.
+    Return:
+        args, which containers parsed input arguments.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--framework",
+                        choices=["tensorflow", "mxnet", "pytorch"],
+                        help="framework of container image.",
+                        default="N/A",
```
Comment (arjkesh) [inline deep_learning_container.py]: maybe sync up with the BI team on what a good default for this is?

**Comment by arjkesh** [comment]
Note that there is an associated unit test for this file (https://github.com/aws/sagemaker-tensorflow-container/blob/master/test/unit/test_deep_learning_container.py) -- you might consider maintaining it here and running it as part of your changes to the file to ensure that you're not introducing any breaking changes

File: deep_learning_container.py (hunk 3 above)
Comment (TusharKanekiDey) [reply deep_learning_container.py]: I made the change in current PR.

File: deep_learning_container.py (hunk 2 above)
Comment (TusharKanekiDey) [reply deep_learning_container.py]: I made the change in current PR.

File: deep_learning_container.py (hunk 4 above)
Comment (TusharKanekiDey) [reply deep_learning_container.py]: I have gone with the first option in current PR. Will discuss with BI team once as well

File: deep_learning_container.py (hunk 5 above)
Comment (TusharKanekiDey) [reply deep_learning_container.py]: I made the change in current PR.

File: deep_learning_container.py (hunk 6 above)
Comment (TusharKanekiDey) [reply deep_learning_container.py]: I have messaged the BI regarding this.

**Comment by TusharKanekiDey** [comment]
I have made a separate PR as we discussed for adding unit test(https://github.com/aws/aws-deep-learning-containers-utils/pull/3).

File: deep_learning_container.py [hunk 7]
```diff
@@ -69,18 +69,51 @@ def _retrieve_instance_region():
...
     instance_id = _retrieve_instance_id()
     region = _retrieve_instance_region()
+    args = parse_args()
+    framework, framework_version, job = args.framework, args.framework_version, args.job
+    py_version = sys.version.split(" ")[0]
 
     if instance_id is not None and region is not None:
         url = ("https://aws-deep-learning-containers-{0}.s3.{0}.amazonaws.com"
-               "/dlc-containers.txt?x-instance-id={1}".format(region, instance_id))
+               "/dlc-containers.txt?x-instance-id={1},framework={2},framework_version={3},py_version={4},job={5}".format(region, instance_id, framework, framework_version, py_version, job))
```
Comment (arjkesh) [inline deep_learning_container.py:112]: right now we will be sending requests with framework=N/A...etc if none of the arguments are set.  I think we should have a default like None, and then conditionally add the param to the request if it is not none.

i.e.

```
additional_params = ''
if framework:
    additional_params+=",framework={framework}"
...
```

Comment (TusharKanekiDey) [reply deep_learning_container.py:112]: Okay, this looks a much better approach. I will add this in the code then.

Comment (TusharKanekiDey) [reply deep_learning_container.py]: As we talked, made all the arguments required instead of optional

File: deep_learning_container.py [hunk 8]
```diff
@@ -69,18 +69,48 @@ def _retrieve_instance_region():
...
     region = _retrieve_instance_region()
+    args = parse_args()
+    framework, framework_version, job = args.framework, args.framework_version, args.job
+    py_version = sys.version.split(" ")[0]
 
     if instance_id is not None and region is not None:
         url = ("https://aws-deep-learning-containers-{0}.s3.{0}.amazonaws.com"
                "/dlc-containers.txt?x-instance-id={1}".format(region, instance_id))
         response = requests_helper(url, timeout=0.2)
+        if (os.environ.get('TEST_MODE') == str(1)):
```
Comment (arjkesh) [inline deep_learning_container.py:111]: not seeing any diff for the url string above, does this neeed to be re-added?

Comment (TusharKanekiDey) [reply deep_learning_container.py:111]: Sorry, my bad.

File: deep_learning_container.py [hunk 9]
```diff
@@ -69,18 +69,48 @@ def _retrieve_instance_region():
...
     instance_id = _retrieve_instance_id()
     region = _retrieve_instance_region()
+    args = parse_args()
+    framework, framework_version, container_type = args.framework, args.framework_version, args.container_type
+    py_version = sys.version.split(" ")[0]
 
     if instance_id is not None and region is not None:
         url = ("https://aws-deep-learning-containers-{0}.s3.{0}.amazonaws.com"
-               "/dlc-containers.txt?x-instance-id={1}".format(region, instance_id))
+               "/dlc-containers.txt?x-instance-id={1},framework={2},framework_version={3},py_version={4},container_type={5}".format(region, instance_id, framework, framework_version, py_version, container_type))
```
Comment (arjkesh) [inline deep_learning_container.py:109]: nit: can you make this line less long? i think it is well over 120 characters

in the new ci/cd repo we've been using a formatting tool called black, which you can use pretty easily.

just pip install black, then run `black -l 120 /path/to/file`

File: deep_learning_container.py [hunk 10]
```diff
@@ -52,35 +52,82 @@ def _retrieve_instance_region():
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--framework", choices=["tensorflow", "mxnet", "pytorch"], help="framework of container image.", required=True
+    )
+    parser.add_argument("--framework-version", help="framework version of container image.", required=True)
+    parser.add_argument(
+        "--container-type",
+        choices=["training", "inference"],
+        help="What kind of jobs you want to run on container. \
+                                  Either training or inference.",
```
Comment (arjkesh) [inline deep_learning_container.py:100]: not sure what happened here, but this can be all on the same line right?

Comment (TusharKanekiDey) [reply deep_learning_container.py:100]: Yes , it is better to have it in one line

**Pull request merged**